### PR TITLE
Adding Glean back to the iOS megazord

### PR DESCRIPTION
**Pull request by travis79**
This is so we can use it to record metrics in Rust on iOS

### CODE REVIEW

**Comment by travis79** [comment]
Seems I forgot to move the exposure event to Rust completely for iOS...   not quite ready yet.

File: components/nimbus/src/lib.rs
Comment (jhugman) [inline components/nimbus/src/lib.rs:506]: I feel sure we can do something a bit more efficient than this, but I can't think of a better way that's as non-invasive as this.

File: components/nimbus/src/nimbus.udl
Comment (jhugman) [inline components/nimbus/src/nimbus.udl:206]: 😻

**Review by jhugman — APPROVED** [review]
This is looking lovely. 😻 

I have added comments which are mostly filing bugs and questions for my own understanding. 

I haven't reviewed anything in xcodeproj files, or .h files, and so we should probably ask one of the iOS teams to look at those and give a r+.

From my point of view, this is brilliant, it's been a long time coming, so thank you.

File: components/nimbus/src/enrollment.rs
Comment (travis79) [reply components/nimbus/src/enrollment.rs:445]: Yes, we should no longer need this. Unfortunately we rely upon this structure in our tests so removing it completely is a bit messy.

File: components/nimbus/ios/Nimbus/Nimbus.swift
Comment (travis79) [reply components/nimbus/ios/Nimbus/Nimbus.swift:8]: Same reason as before, Nimbus can't find Glean in the megazord without this. I don't know if MozillaRustComponents is always available or not, just following the pattern seen in other components that have dependencies inside of AppServices

File: components/nimbus/src/lib.rs
Comment (travis79) [reply components/nimbus/src/lib.rs:506]: Yes, hoping @tarikeshaq might have some ideas on this. It was as close to the logic of what we were doing in Swift as I could arrive at.

File: components/nimbus/tests/test_telemetry.rs
Comment (travis79) [reply components/nimbus/tests/test_telemetry.rs:121]: There should already be an issue I think, around removing the Kotlin and Swift Glean instrumentations. Let me see if I can find and link it.

File: .circleci/config.yml
Comment (tarikeshaq) [inline .circleci/config.yml:123]: Do you know if we need to include this in our `building` docs: https://github.com/mozilla/application-services/blob/main/docs/building.md?

File: components/nimbus/Cargo.toml
Comment (tarikeshaq) [inline components/nimbus/Cargo.toml:49]: I don't think we need to include it in the dev-dependencies if it's already a dependency

File: components/nimbus/Cargo.toml
Comment (tarikeshaq) [inline components/nimbus/Cargo.toml:38]: Hmmm I'll need to think about this a little, from an "impact on android" perspective.

Two questions:
1) is always guaranteed that if the glean binary isn't initialized, all glean API's are no-ops? Because that will be the case here in android, where they use libxul
2) How OK are we with duplicating the glean binary in android?

I'm not sure about the first question, and the second question we can maybe measure?

Alternatively, for now, we can gate including glean on a feature https://doc.rust-lang.org/cargo/reference/features.html

we can even have it enabled by default, then in the [megazord for android](https://github.com/mozilla/application-services/blob/main/megazords/full/Cargo.toml) we can turn the feature off

File: components/nimbus/build.rs
Comment (tarikeshaq) [inline components/nimbus/build.rs:11]: Hmmm will this always be true? Like, if a mac user follows our instructions in the building docs, will they end up with a `python3` binary?

File: components/nimbus/build.rs
Comment (tarikeshaq) [inline components/nimbus/build.rs:32]: Is there a way to print some more information here in case a build fails for a developer? Something that can point them in some direction

File: components/nimbus/ios/Nimbus/GleanPlumbHelpers.swift
Comment (tarikeshaq) [reply components/nimbus/ios/Nimbus/GleanPlumbHelpers.swift:8]: The `#if canImport`... was added when we had our code used in both a Carthage and a Swift Package Manager environment. The `MozillaRustComponents` wasn't used in Firefox iOS at the time, but was used in Focus.

Now that Carthage is out of the way, we can probably get rid of the `#if canImport` and just import `MozillaRustComponents` - we can file a follow-up to clean it up from everywhere.


All that said, we probably still want to keep this in any Glean files that are added, assuming glean's local development flow and local testing doesn't build our megazord

File: components/nimbus/src/enrollment.rs
Comment (tarikeshaq) [inline components/nimbus/src/enrollment.rs:381]: nit: do we need the `clone` here? I would of thought `to_string()` is enough as that probably allocates a new string anyways

(and ditto for below)

Comment (tarikeshaq) [reply components/nimbus/src/enrollment.rs:381]: As a tiny follow-up we file, I notice we have to clone a bunch of times, when the `branch` and `enrollment_id` values don't really change

but that makes sense since the generated glean metrics code expects `String`s... can we file a ticket/issue against glean/glean-metrics to allow for strings to be reference, or just use a [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html)

File: components/nimbus/src/lib.rs
Comment (tarikeshaq) [reply components/nimbus/src/lib.rs:506]: Yeah I don't have any brilliant ideas here 😞 We can modify our state to include a map from feature_id -> experiment maybe? but that's a little intrusive and we'll need to be careful keeping it up to date as experiments change

Regardless I'm happy with this, and we can always improve it if we recognize that the time spent iterating over experiments is too expensive

File: megazords/ios-rust/Cargo.toml
Comment (tarikeshaq) [inline megazords/ios-rust/Cargo.toml:24]: out of curiosity, why do we need to include both? do we need to expose the `rlb` at this layer?

File: megazords/ios-rust/build-xcframework.sh
Comment (tarikeshaq) [inline megazords/ios-rust/build-xcframework.sh:149]: hmmm we'll need to be extra careful here to make sure the `uniffi_bindgen` version glean uses to generate its scaffolding is compatible with the version we use here. This also means that we are essentially tying our `uniffi` version in `application-services` with the `uniffi` version used in glean (at least in the short term, until we get the `uniffi` versioning work done.

Versions being tied doesn't sit well with me, but I can't see a way around it for now/

So the consequences of this are:
-  Whenever we bump the glean submodule (and it includes an updated `uniffi`) the iOS builds will fail, and CI will catch them forcing us to either align the version here in application-services, or go back and align the version in glean and cutting another release there
- We will **not** be able to bump uniffi in this repository on its own. Whenever we want to bump uniffi here, we'll have to first bump it in glean, cut a release there, then update it here

I'll try to spend a bit of time thinking about this so we can move without blocking this work
cc @mhammond and @bendk

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Thanks a ton @travis79 🚀 🎉 !! Once we get this through we should have a good story for iOS

I had a couple of blocking thoughts/comments that we might need to work out before we land this:

1) The comment about impact on android and possibly feature gating
2) The uniffi version compatibility comment

The rest are nits and questions. I'll timebox sometime to think about the `uniffi` comment and see if we can comeup with a good way to move forward!

File: components/nimbus/src/lib.rs
Comment (jhugman) [reply components/nimbus/src/lib.rs:506]: > We can modify our state to include a map from feature_id -> experiment maybe? but that's a little intrusive and we'll need to be careful keeping it up to date as experiments change
Yes, we do keep a map of `feature_id` to `EnrolledFeatureConfig`, which could keep an optional experiment id.

It's kept in memory and updated everytime `apply_pending_experiments` is called.

> Regardless I'm happy with this, and we can always improve it if we recognize that the time spent iterating over experiments is too expensive

Agreed.

File: megazords/ios-rust/build-xcframework.sh
Comment (tarikeshaq) [reply megazords/ios-rust/build-xcframework.sh:149]: I wanted to point out that I chatted briefly with @bendk about this, and with the upcoming changes in https://github.com/mozilla/uniffi-rs/pull/1271 I'm less concerned about the coupling between the versions. That PR changes compatibility to be less about the uniffi version and more about the low-level contract between the bindings and the scaffolding (which we don't change often at all)

And since the UDL we use is the same one from the submodule, there shouldn't be any incompatibility there.

I won't block on this anymore, and even if this PR lands first - the uniffi naming changes are already approved/agreed upon so it'll catch up quickly 🚀

File: .circleci/config.yml
Comment (travis79) [reply .circleci/config.yml:123]: We shouldn't, this just isn't installed by default with python in the CI environment. Typical local building you would get pip along with python3 installation (or a venv)

File: megazords/ios-rust/Cargo.toml
Comment (travis79) [reply megazords/ios-rust/Cargo.toml:24]: You may be correct here, I just added both, but I can try without the rlb dependency here.

File: components/nimbus/Cargo.toml
Comment (travis79) [reply components/nimbus/Cargo.toml:38]: The answer to the first question is yes, uninitialized Glean is a no-op except for the queue for metrics recorded prior to init, but once it is full it too becomes a no-op. I don't expect much overhead from Glean in this regard and I am confident in Glean's logic for this.

Adding the RLB is just adding the binding layer on top of glean-core, so it's a fairly thin shim. There should still only be one glean-core coming from libxul and the rlb will interact with it on Android, and interact with the megazorded glean-core on iOS.  I didn't notice any big size regressions while testing in Focus, but happy to do a more detailed measurement if product has any issues with it.

File: components/nimbus/build.rs
Comment (travis79) [reply components/nimbus/build.rs:11]: Jan-Erik has created a glean-build crate as a wrapper around the glean_parser step, so this script should no longer be needed. I'll be pushing a commit to migrate to this build crate to address your concern here, this was simply a hack to get it to work with Windows development.

File: components/nimbus/build.rs
Comment (travis79) [reply components/nimbus/build.rs:32]: The glean_parser output would be the output that contains the information needed. Again, with my plan to switch to the glean-build crate rather than the script, the output should be available. If it isn't, I'll file a bug to see that the important info makes it into stdout.

File: components/nimbus/src/enrollment.rs
Comment (travis79) [reply components/nimbus/src/enrollment.rs:381]: I am probably a little over-zealous with `clone()` as a way to get along with the borrow checker, I'll try your suggestions to see if I can make this better, but at least these aren't large objects so I wasn't too worried about cloning them.

File: megazords/ios-rust/build-xcframework.sh
Comment (travis79) [reply megazords/ios-rust/build-xcframework.sh:149]: I think A-S and Glean are likely going to both be using a fairly recent version of uniffi most of the time, so hopefully the compatibility problems will be minimal. Since there are already changes proposed that will make this even easier, I'm okay with keeping Glean on the same version of A-S, and I'm sure we can work it out if we need to differ for some reason.

File: components/nimbus/ios/Nimbus/Nimbus.swift
Comment (travis79) [reply components/nimbus/ios/Nimbus/Nimbus.swift:112]: [Issue Filed](https://mozilla-hub.atlassian.net/browse/EXP-2630)

File: components/nimbus/Cargo.toml
Comment (tarikeshaq) [reply components/nimbus/Cargo.toml:38]: Ideally, I still would like to see either a [`cfg` on android](https://doc.rust-lang.org/reference/conditional-compilation.html#target_os) or some type of cargo feature that protects non-ios consumers from including a duplicate glean (and helps if anyone is confused why calls are no-ops) - if that does seem like a lot of work for what we would remove when glean-in-rust hits android I'm willing to concede, I'll leave it as a nice to have as long as android still builds and works OK

File: components/nimbus/build.rs
Comment (tarikeshaq) [inline components/nimbus/build.rs:11]: I love this!!! A lot better than shelling out 😍 🎉

File: components/nimbus/tests/test_telemetry.rs
Comment (tarikeshaq) [inline components/nimbus/tests/test_telemetry.rs:14]: nit: do we need this? would the `cfg` on the `test` module protect against this already?

File: megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/LoginsTests.swift
Comment (tarikeshaq) [inline megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/LoginsTests.swift:15]: Just out of curiosity, what's going in here and why can we get rid of it (instead of calling some other function more directly)

File: megazords/ios-rust/MozillaTestServices/glean.udl
Comment (tarikeshaq) [inline megazords/ios-rust/MozillaTestServices/glean.udl:1]: Curious why we needed to add a second `udl` here - is this the same `udl` in glean? Is it possible to point to the submodule in all of our scripts instead of duplicating the UDL?

File: megazords/ios-rust/ios/Generated/glean.swift
Comment (tarikeshaq) [inline megazords/ios-rust/ios/Generated/glean.swift:1]: Why do we need this file? Shouldn't this be generated by `uniffi` on `rust-components-swift`? I don't mind it generated here, but we'll need some consistent script that runs to generated updates.


(if we need it here, we can start a discussion around generally checking-in generated uniffi files, which TBH I'm in support of, but it seems beyond the scope of this PR)

File: megazords/ios-rust/ios/Generated/gleanFFI.modulemap
Comment (tarikeshaq) [inline megazords/ios-rust/ios/Generated/gleanFFI.modulemap:1]: ditto on generated file, (and I guess the `.h` file too)

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Almost there! Thanks @travis79!!!

Once you've addressed the comments I've dropped we should be in a good place to land this.

The last thing I want to confirm before we land this, can we try a local build of Fenix with this branch to confirm that we aren't causing any breakages. It would also be ideal if we can confirm that Fenix is still recording telemetry as expected and that nimbus is working properly

File: megazords/ios-rust/MozillaTestServices/glean.udl
Comment (travis79) [reply megazords/ios-rust/MozillaTestServices/glean.udl:1]: Yes, that was my intention, Glean now has two udl files and does some swapping between them when building the iOS megazord (via somewhat hacky use of file renaming, but it works for now). I think I just forgot to remove this one once I moved the udl needed to build the iOS megazord compatible Glean into the Glean repo.

File: megazords/ios-rust/ios/Generated/glean.swift
Comment (travis79) [reply megazords/ios-rust/ios/Generated/glean.swift:1]: This shouldn't be here, I think it's a side effect of me playing around with the TestServices project trying to get it to work.  Let me clean that up...

File: megazords/ios-rust/ios/Generated/gleanFFI.modulemap
Comment (travis79) [reply megazords/ios-rust/ios/Generated/gleanFFI.modulemap:1]: Ditto on this shouldn't be here and may just be leftover from playing with getting the TestServices to build

File: components/nimbus/tests/test_telemetry.rs
Comment (travis79) [reply components/nimbus/tests/test_telemetry.rs:14]: I think it might, this is most likely just some copy-pasta from Glean test examples.

File: megazords/ios-rust/build-xcframework.sh
Comment (travis79) [reply megazords/ios-rust/build-xcframework.sh:149]: No, that udl in TestServices isn't supposed to be there so we should rely on the one in the submodule

File: megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/LoginsTests.swift
Comment (travis79) [reply megazords/ios-rust/MozillaTestServices/MozillaTestServicesTests/LoginsTests.swift:15]: Which probably means I can remove the import statement too :D

File: components/nimbus/src/enrollment.rs
Comment (tarikeshaq) [inline components/nimbus/src/enrollment.rs:395]: super nit: we can `cfg(feature=` a block here instead of multiple `cfg`s, i.e adding curly brackets before and after the part that is for glean-in-rust

File: megazords/ios-rust/Cargo.toml
Comment (tarikeshaq) [inline megazords/ios-rust/Cargo.toml:12]: At first I was going to suggest replacing this with a `features = ["builtin-glean"]` entry in the dependencies section for nimbus, but I realized that I actually like this - in case we ever want to test the not-glean-in-rust in iOS! Awesome! 🚀

**Review by tarikeshaq — APPROVED** [review]
🚀 🚀 AWESOME!!! Thanks a ton for driving this @travis79


LETS 🚢 :shipit:

File: components/nimbus/src/enrollment.rs
Comment (tarikeshaq) [reply components/nimbus/src/enrollment.rs:395]: eeeh I relalized there is a line in between that is not cfg'ed nvm this comment!

Comment (travis79) [reply components/nimbus/src/enrollment.rs:395]: Yeah, I should have done that here, the extra isn't really used without Glean and I missed conditionally compiling it away anyways.  I'll update this to use the `cfg!() {}` macro and wrap the whole block of Glean related code.

**Comment by travis79** [comment]
Adding the `do-not-merge` label until we get ready to land this for v107 nightlies (mid-September)

**Comment by tarikeshaq** [comment]
@travis79 since this has gotten a little stale, do you know what the priority of this work is now? (i.e do we expect we'll get to it anytime soon?)

I don't object to keeping the PR open until we get to it, just curious what a timeline might look like if we have one?

**Comment by travis79** [comment]
> 

@tarikeshaq Yes, the priorities on this shifted a bit. This is still desirable to do but I don't have time to get back to it immediately. Ultimately I'd love to be able to solve both iOS and Android when we get back to this. Let's close this in favor of starting fresh without any bitrot to deal with.

**Pull request closed without merging**
